pmlc/ast/builder.cc

```diff
@@ -740,6 +741,40 @@ struct ProgramBuilder {
         .result();
   }
 
+  Value makeGatherOp(ExprNodeIntrinsic *node, ArrayRef<Value> operands) {
+    TensorShape shape = evaluator.getShape(node);
+    RankedTensorType resultType = builder.getRankedTensorType(shape);
+    IntegerAttr axis;
+    IntegerAttr interpolationMode;
+    IntegerAttr nearestMode;
+    FloatAttr cubeCoeff;
+    if (!matchPattern(operands[2], m_Constant(&axis))) {
+      throw std::runtime_error("'gather' primitive expects the 'axis' argument "
+                               "to be a constant integer");
+    }
+    if (!matchPattern(operands[3], m_Constant(&interpolationMode))) {
+      throw std::runtime_error(
+          "'gather' primitive expects the 'interpolationMode' argument "
+          "to be a constant integer");
+    }
+    if (!matchPattern(operands[4], m_Constant(&nearestMode))) {
+      throw std::runtime_error(
+          "'gather' primitive expects the 'nearestMode' argument "
+          "to be a constant integer");
+    }
+    if (!matchPattern(operands[5], m_Constant(&cubeCoeff))) {
+      throw std::runtime_error(
+          "'gather' primitive expects the 'cubeCoeff' argument "
+          "to be a constant float");
+    }
+    auto cubeCoeffFloat = static_cast<float>(cubeCoeff.getValueAsDouble());
```

Comment:
Why is this cast necessary? We should store constants in the highest precision possible.

Reply:
Fixed, set cubeCoeff as f64 attributes and added support for f16, f32 ,f64. f16 works on CPU. MLIR spirv not fully support f16 right now.